src/engine/SPARQLProtocol.h

```diff
@@ -205,5 +105,190 @@ class SPARQLProtocol {
     return tokenFromAuthorizationHeader
                ? std::move(tokenFromAuthorizationHeader)
                : std::move(tokenFromParameter);
-  };
+  }
+};
+
+class SPARQLProtocol {
+  FRIEND_TEST(SPARQLProtocolTest, parseGET);
+  FRIEND_TEST(SPARQLProtocolTest, parseUrlencodedPOST);
+  FRIEND_TEST(SPARQLProtocolTest, parseQueryPOST);
+  FRIEND_TEST(SPARQLProtocolTest, parseUpdatePOST);
+  FRIEND_TEST(SPARQLProtocolTest, parsePOST);
+
+  static constexpr std::string_view contentTypeUrlEncoded =
+      "application/x-www-form-urlencoded";
+  static constexpr std::string_view contentTypeSparqlQuery =
+      "application/sparql-query";
+  static constexpr std::string_view contentTypeSparqlUpdate =
+      "application/sparql-update";
+
+  static ad_utility::url_parser::ParsedRequest parseGET(
+      const ad_utility::httpUtils::HttpRequest auto& request) {
+    using namespace ad_utility::url_parser::sparqlOperation;
+    using namespace ad_utility::use_type_identity;
+    auto parsedRequestBuilder = ParsedRequestBuilder(request);
+    parsedRequestBuilder.extractAccessToken(request);
+
+    const bool isQuery = parsedRequestBuilder.parametersContain("query");
+    if (parsedRequestBuilder.parametersContain("update")) {
+      throw std::runtime_error("SPARQL Update is not allowed as GET request.");
+    }
+    if (parsedRequestBuilder.isGraphStoreOperation()) {
+      if (isQuery) {
+        throw std::runtime_error(
+            R"(Request contains parameters for both a SPARQL Query ("query") and a Graph Store Protocol operation ("graph" or "default").)");
+      }
+      // SPARQL Graph Store HTTP Protocol with indirect graph identification
+      parsedRequestBuilder.extractGraphStoreOperation();
+    } else if (isQuery) {
+      // SPARQL Query
+      parsedRequestBuilder.extractOperationIfSpecified(ti<Query>, "query");
+      parsedRequestBuilder.extractDatasetClauses();
+    }
+    return std::move(parsedRequestBuilder).build();
+  }
+
+  static ad_utility::url_parser::ParsedRequest parseUrlencodedPOST(
+      const ad_utility::httpUtils::HttpRequest auto& request) {
+    using namespace ad_utility::url_parser::sparqlOperation;
+    using namespace ad_utility::use_type_identity;
+    namespace http = boost::beast::http;
+    auto parsedRequestBuilder = ParsedRequestBuilder(request);
```

Comment:
Maybe you can have a dispatcher that first (for any request) constructs the `ParsedRequestBuilder` and extracts the access token, and then all the different code for GET POST and whatever can be moved to the .cpp file.

Reply:
In most cases the parameters come from the URL as the http query parameters. For urlencoded requests though they are read from the body. The access token can only be read once we have the right parameters set.